동기화 메모장[Step2] Hosinging

Hosinging opened the pull request:
안녕하세요 쿠마~! @leejun6694 
2단계를 생각했던대로(코어데이터에 저장된 데이터가 없으면 샘플데이터를 보여주고, 있으면 코어데이터에 저장된 데이터를 보여주고... 등) 적용하는 것이 그리 오래 걸리지않는 작업일 줄 알고 차근차근 필요한 내용들을 공부하면서 진행해보려고 했더니... 공부보다 적용과정에서 놓쳤거나 생각치 못했던 부분들에서 에러가 나와서(데이터 표시해줄 때, 데이터 삭제했을 때..등)해당 부분들에  대한 처리를 최대한 하고 보내느라고 많이 늦어졌습니다. 

---
# 참고
> ### 1단계 merge 전 회의실에서 나눈 대화 내용

말씀 주신 내용에 대한 부분들은 2단계 시작할 때 반영해둔 상태로 시작했습니다. 
예: 스플릿뷰 컨트롤러의 자식뷰들 중에 첫번째인 네비게이션 컨트롤러의 자식뷰들의 첫번째 
```swift
 if let navigationController = splitViewController?.viewControllers.first,
    let memoListTableViewController = navigationController.children.first as? MemoListTableViewController {
     memoListTableViewController.delegate = self
}
```

# 해결하지 못한점
> ### 메모 목록에서 마지막 메모 삭제

MemoListTableViewController에서 샘플데이터의 경우 마지막 남은 데이터까지 제거가 됐을 때 에러가 발생하지 않았지만, 조건식 때문인지 CoreData에 추가하고 삭제하는 경우 마지막 1개까지는 괜찮지만, 그 마지막 한 개를 삭제하고자 했을 때에 똑같은 indexPath.row를 줬지만 index에러가 나는 상황이 있었습니다. 이 부분도 금방 해결할 줄 알고 작업하다가 시간이 길어져서 일단, 이 부분은 패스하고 다른 부분들을 작업했습니다.

> ### ContentViewController 에서 저장 후 바로 삭제, pop 

ContentViewController에서 메모를 작성하고, 저장 완료됨을 확인하고 MemoListTableViewController로 갔다가 다시 들어와서 더보기를 눌러 삭제의 과정을 거치면 실제로 삭제가 되고, 해당 뷰가 pop됨을 확인했습니다. 
근데, 혹시나 싶어서 메모가 저장됐음을 확인하고 바로 더보기 버튼을 눌러서 삭제의 과정을 거치니 해당 뷰가 pop되지 않을 뿐더러, MemoListTableViewController로 갔을 경우 데이터가 그대로 있음을 확인했습니다. 이 부분에 대한 해결을 찾고 해결을 하려면 시간이 더 걸릴 것 같아서 일단 패스하고 다른 부분을 작업했습니다.

> ### IPad에서 popover처리

해당 부분은 인지했고, 공부 과정에서 어떻게 처리해야하는지는 확인을 해둔 상태입니다. 가능하다면 리뷰 반영하는 과정에서 진행을 해봐도 될까요?

---
# 고민한점 & 궁금한점
> ### 1. 자잘자잘한 Extension 추가

많이 사용되진 않지만 '뭔가 있었으면 좋겠다', ' 만들어보는게 도움이 되겠다'는 생각에 파일에 해당 타입(UIView, UIViewController)을 확장하는 메서드가 하나뿐이긴 하지만 필요로 하는 클래스에 만들까 하다가 따로 빼서 만들어봤습니다. 의도와 생각은 이랬지만, 현업에서는 이런 경우 필요로 하는 파일에 바로 작업을 하는지, 이와 같이 extension을 활용하는지 아니면 protocol 또는 다른 방법을 활용하는지 궁금합니다.

> ### 2. 노티피케이션 추가

1번의 질문에서 이어지는 내용인데요. 따로 분리해서 만들어줄까 고민하다가 이 부분은 타입의 확장이 아닌 필요로 하는 클래스를 확장하는 패턴으로 만들어서 사용해봤습니다.

Hosinging committed:
refactor: 텍스트뷰 생성 및 기기에 따라 폰트 사이즈 다르게 표시하는 코드 함수에서 클로저 실행 코드로 변경

Hosinging committed:
chore: self키워드 없어도 되는 곳은 제거.

Hosinging committed:
feat: 메모Entity 및 attribute생성

Hosinging committed:
feat: DataManager클래스 싱글턴으로 생성

Hosinging committed:
feat: DataManager 클래스 내 코어데이터 스택 초기화 코드, 메모 CRUD코드 구현

Hosinging committed:
chore: DataManager.swift파일 내 NSPersitentContainer를 상속받는 NSPersitentCloudKitContainer로 되돌림

Hosinging committed:
chore: 불필요 코드 제거

Hosinging committed:
feat: ContentViewController.swift파일 내 화면 진입 시 텍스트뷰 관련 동작 필요한 메서드 추가
- 키보드가 보여질 때 키보드가 텍스트뷰를 덮어 내용이 가려지지 않게해주고 스크롤도 그에 맞게 조절되게 해주는 scrollWhenKeyboardWillAppear()메서드 추가
- 뷰가 로드 됐을 때 컨텐츠의 상단이 보이게끔 contenOffSet조정하는 scrollWhenContentTextViewDidAppear()메서드 추가

Hosinging committed:
chore: ContentViewController.swift 파일 내 contentTextView contentInset관련 로컬 변수 명 수정

Hosinging committed:
refactor: DataManager.swift파일 내 코어데이터에 CRUD하는 메서드 일부 수정
- extension으로 메서드 위치 이동
- 파라미터 아규먼트레이블 지정

Hosinging committed:
chore: DataManager.swift파일 내 모델 이름 수정

Hosinging committed:
feat: UIAlertController+addAction.swift 파일 내 코드 추가
- UIAlertController에 UIAlertAction을 배열로 추가해주는 func addActions(_:)추가

Hosinging committed:
feat: ContentViewController.swift파일 내 코드 추가
- navigationBarButtonItem을 눌렀을 때 나오는 액션시트를 생성하고, 액션을 추가하는 createMoreDetailSheet()메서드 추가
- rightBarButtonItem에 action으로 사용할 showSheetMoreDetail() @objc 메서드 추가

Hosinging committed:
feat: MemoListTableViewController.swift 파일 내 코드 추가
- tableView(_:canEditRowAt:) 메서드 추가
- tableView(_:editingStyleForRowAt:) 메서드 추가
- tableView(_:commit:indexPath:) 메서드 추가

Hosinging committed:
chore: CoreData Entity의 Attribute오타 수정하는 과정에서 불필요한 파일 제거

Hosinging committed:
feat: UIViewController에서 사용가능한 showAlert(title:message:)메서드 추가

Hosinging committed:
chore: CoreData Entity의 Attribute 오타 수정으로 인한 수정 필요 부분 처리

Hosinging committed:
feat: ContentViewController.swift파일 내 코드 추가
- textView 리턴키 타입 지정
- textView 델리게이트 지정 및 textView(_:shouldChangeTextIn:replacementText:)구현
- textView내 값이 업데이트 되거나, 값 저장할 때 필요한 Notification Extension에 추가

Hosinging committed:
feat: MemoListTableViewController에 코드 추가
- ContentViewController의 텍스트뷰에서 값이 저장됐을 때 받을 노티피케이션 옵저버 추가 및 옵저버 해제 코드 추가
- viewWillAppear(_:)메서드 내에서 코어데이터에 저장된 메모를 가져오고, 테이블 뷰를 리로드해주는 코드 추가
- tableView(_:numberOfRowsInSection:), tableView(_:cellForRowAt:indexPath:)메서드 내에 코어데이터 저장된 메모의 카운트에 따른 테이블뷰 셀에 분기 표시 코드 구현

Hosinging committed:
# <타입>: <제목>
test: 테스트 커밋
##### 제목은 최대 50 글자까지만 입력 ############## -> |
주석들어간게 보일까
# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# 꼬릿말은 아래에 작성: ex) #이슈 번호

# --- COMMIT END ---
# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# ------------------
#     제목 첫 글자를 대문자로
#     제목은 명령문으로
#     제목 끝에 마침표(.) 금지
#     제목과 본문을 한 줄 띄워 분리하기
#     본문은 "어떻게" 보다 "무엇을", "왜"를 설명한다.
#     본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
# ------------------

Hosinging committed:
Revert "# <타입>: <제목>"

This reverts commit eb6aeb4ee20734a887b46f55bdfb1857130ef346.

Hosinging committed:
feat: ContentViewController.swift파일 내 코드 추가
- 텍스트뷰에 기존 메모인지, 새 메모인지에 따라 표시 분기해주는 메서드 setMemoIfNewMemoOrOriginalMemo()추가
- 텍스트뷰 델리게이트를 지정해주는 메서드 setMemoListTableViewDelegate()추가

Hosinging committed:
fix: MemoListTableViewController에 코드 수정
- tableView(_:didSelectRowAt:)메서드 내에서 기존 샘플 데이터를 넘겨주던 부분을 코어데이터 엔티티를 전달해주는 형태로 수정

Hosinging committed:
feat: ContentViewController.swift 파일 내 코드 추가
- viewWillAppear(_:)메서드 내 코어데이터 저장 메모 유무에 따라 분기 표시 코드 추가
- 취소, 삭제 버튼있는 alert띄워주는 showAlertContainTowAction(title:message:)메서드 추가
- 메모가 삭제 됐을 때 보낼 memoDidDelete notification선언

Hosinging committed:
feat: MemoListTableViewController.swift 파일 내 코드 추가
- 메모가 코어데이터에 저장할 때, 삭제될 때 보내지는 notification을 받을 옵저버 관련 메서드 addObserverToMemoInsert(), addObserverToMemoDelete 추가
- tableView(_:didSelectRowAt:indexPath:)메서드내 코어데이터 저장된 메모 갯수에 따라 샘플데이터와 분기해서 표시하는 코드 추가
- tableView(_:commit:forRowAt:)메서드 내 코어데이터 저장된 메모 갯수에 따라 샘플데이터와 분기해서 삭제하는 코드 추가

Hosinging committed:
chore: 기존 작성코드로 복구

Hosinging committed:
refactor: 중복되고, 가독성 떨어지는 UINavigationController접근 코드 수정

Hosinging committed:
chore: 불필요한 코드 제거 및 전체 코드 줄맞춤 처리

Hosinging committed:
chore: 줄맞춤

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
+            textView.font = .systemFont(ofSize: 20, weight: .bold)
+        } else {
+            textView.font = .systemFont(ofSize: 25, weight: .bold)
+        }
```
아래와 같은 코드로도 구현이 가능합니다. 참고해주세요~~!
```
let fontSize = UITraitCollection.current.userInterfaceIdiom == .phone ? 20.0 : 25.0
textView.font = .systemFont(ofSize: fontSize, weight: .bold)
```

Hosinging replied:
삼항 연산자의 적용이 가능한 코드였는데 생각못했었네요~! 해당 내용으로 코드 적용을 해봤습니다.

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
+        setMemoIfNewMemoOrOriginalMemo()
+
+        setMemoListTableViewDelegate()
+        contentTextView.delegate = self
```
`contentTextView` 를 생성하는 곳에 코드를 작성하면 어떨까요?
https://github.com/yagom-academy/ios-cloud-notes/pull/67/files#diff-dfb649ba102719cb80ffd656ef03e07598e938b5a34d19864336b1e246196bbcR11

Hosinging replied:
> 접근1

해당 클로저 내부에서는 ContentViewController의 인스턴스를 self로 접근이 안되는 것으로 알고있었지만, 접근하려고 해봤더니 아래와 같이 할당할 수 없다는 에러가 나왔었습니다.

```swift
textView.delegate = self
```

#### Cannot assign value of type '(ContentViewController) -> () -> ContentViewController' to type 'UITextViewDelegate?'

> 접근2

 self가 해당 타입의 인스턴스로 알고 있어서 인스턴스를 만들어서 넣어주려고 해봤더니 아래와 같은 경고 문구가 나왔었습니다.

#### Instance will be immediately deallocated because property 'delegate' is 'weak'


> 접근3 
delegate지정 객체가 초기화가 안된 시점에 self의 접근이 안되는게 당연한 것 같아서, delegate지정 객체가 초기화되고, textView에 접근하는 시점에 self로 접근이 가능할 것 같아서, 접근할 때 초기화가 되는 lazy키워드를 사용해봤더니 self에 접근이 가능해졌는데요. 
```swift 
lazy var contentTextView: UITextView = {
     ...
    textView.delegate = self
    return textView
}()
```

> 클로저 실행문 내부에서 특정 객체의 델리게이트로 해당 인스턴스를 할당해줄 때는 이와 같이 lazy키워드를 사용해서 해주는게 맞는지 궁금합니다.

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
+    private func configureTextView() {
+        view.addSubview(contentTextView)
+        contentTextView.setConstraintEqualToAnchor(superView: view)
+
```
불필요한 줄바꿈이 있습니다.

Hosinging replied:
확인했습니다. 🙂

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
+            originalMemoContent = memoEntity.content
         } else {
-            textView().font = UIFont.systemFont(ofSize: 25, weight: .bold)
+            contentTextView.text = ""
```
"" 대신 nil 은 어떨까요?
물론 상관 없지만,, 개인적으로 코드 내 쌍따옴표를 지양하고 있습니다ㅎㅎ

Hosinging replied:
오~! nil을 생각을 못해보고 항상 쓰던대로 빈문자는""로 사용을 했었네요. 
상관이 없지만, 개인적으로 코드 내 쌍따옴표를 지양하시는 이유를 여쭤봐도 될까요?🥲

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
     private func configureNavigationBar() {
         let itemImage = UIImage(systemName: "ellipsis.circle")
-        let rightBarButtonItem = UIBarButtonItem(image: itemImage, style: .plain, target: nil, action: nil)
+        let rightBarButtonItem = UIBarButtonItem(image: itemImage, style: .plain, target: self, action: #selector(showSheetMoreDetail))
```
`item`, `right` 으로 해당 버튼이 어떤 동작을 하고 있고, 역할을 하는지 분명하지 않은 것 같습니다.
네이밍을 좀 더 고려해보면 어떨까요?

Hosinging replied:
itemImage -> seeMoreButtonImage
rightBarButtonItem -> seeMoreBarButtonItem과 같이 좀 더 명확하게 역할을 표현할 수 있는 네이밍으로 수정해봤습니다.

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
+
+    private func createMoreDetailSheet() -> UIAlertController {
+        let sheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+        let shareAction = UIAlertAction(title: "Share", style: .default) { [weak self] _ in
```
https://github.com/yagom-academy/ios-cloud-notes/pull/67/files#diff-dfb649ba102719cb80ffd656ef03e07598e938b5a34d19864336b1e246196bbcR61
위 closure 에선 `weak self` 가 없었는데, 여긴 있네요ㅎㅎ
혹시 어떤 차이점이 있을까요!?

Hosinging replied:
아, 클로저 캡쳐리스트를 공부할 때 클로저 안에서 self.으로 접근하는 코드가 있다면 강한 참조 사이클이 발생할 수 있다고하여, [weak self]를 써주는게 좋다. 라는 내용을 본적이 있어서 weak self를 사용해봤는데요. 위의 내용에서는 해당 내용을 누락했던 것 같습니다. 
일단 처음 의도대로 코드를 수정하고, 해당 내용에 대해서 한번 찾아보고 코드를 다시 살펴보겠습니다.🙂

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
+            self?.present(activityViewController, animated: true, completion: nil)
+        }
+        let deleteAction = UIAlertAction(title: "Delete", style: .destructive) { [weak self] _ in
+
```
불필요한 줄바꿈이 있습니다.

Hosinging replied:
확인했습니다.🙂

leejun6694 commented on CloudNotes/CloudNotes/Controllers/MemoListTableViewController.swift:
```diff
@@ -6,39 +6,59 @@
 
+    // MARK: - Method
+    private func addObserverToMemoInsert() {
+        token = NotificationCenter.default.addObserver(forName: ContentViewController.newMemoDidInput, object: nil, queue: .main, using: { [weak self] _ in
```
코드 한 줄이 길어지는 경우, 적절한 줄바꿈을 해주면 더 좋을 것 같습니다.

Hosinging replied:
'.'을 기준으로 줄바꿈을 하는 코드를 본적이 있습니다. 위의 코드에서는 .을 기준으로 줄바꿈을 해도 에러가 발생하지 않고,  파라미터의 ','를 기준으로 해도 에러가 발생하지 않는 것을 확인했는데요. 쿠마라면 어떤 식으로 줄바꿈을 하실지 궁금합니다~!

leejun6694 commented on CloudNotes/CloudNotes/Controllers/ContentViewController.swift:
```diff
@@ -8,53 +8,173 @@ import UIKit
 }
+
+// MARK: - Notification
+extension ContentViewController {
```
특정 Controller 내 static 변수보단 `Notification.Name` 을 활용해보는건 어떨까요?

leejun6694 replied:
hint : `extension`, `Notification.Name init`

Hosinging replied:
다음과 같이 Notification.Name을 확장하고, enum과 init을 이용해서 반영해봤습니다.
```swift
extension Notification.Name {
    enum Notifications: String {
        case newMemoDidInput
        case memoDidUpdate
        case memoDidDelete
    }

    init(_ notification: Notifications) {
        self = Notification.Name(notification.rawValue)
    }
}
```

leejun6694 commented on CloudNotes/CloudNotes/Controllers/MemoListTableViewController.swift:
```diff
@@ -65,17 +85,31 @@ extension MemoListTableViewController {
 // MARK: - TableView Delegate Method
 extension MemoListTableViewController {
     override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        if DataManager.shared.memoList.count > 0 {
```
`table view` 에 어떤 데이터를 보여줄지 미리 결정을 하면 어떨까요?
미리 결정된다면 `table view data source` 내 코드가 간결해질 것 같습니다. (if 문들이 사라지겠네요ㅎㅎ)

leejun6694 replied:
`memoList` 와 `parsedDatas` 구현의 차이가 있다면 `protocol` 등의 방법으로 맞춰줄 수 있을 것 같습니다.
(혹시 이해가 잘 안된다면 말씀 부탁드려요!)

leejun6694 replied:
`dateFormatter` 도 `DataManager` 가 아닌, 다른 곳에서 정의되겠네요ㅎㅎ

Hosinging replied:
> 1.  '어떤 데이터를 보여줄지 미리 결정을 하면' 이라는 의미가 정확히 어떤 의미인지 이해하지 못했습니다.

coreData에 저장된 메모의 갯수가 0보다 클 경우 코어데이터의 데이터가, 그게 아닐 경우 디코딩된 샘플 데이터의 데이터가 보여지게 구현을 해둔 상태인데 조건식에 따른 분기가 없이 가능한 것인지...모르겠습니다...

> 2. '구현의 차이가 있다면 protocol등의 방법으로 맞춰줄 수 있을 것 같다'는 의미가 정확히 어떤 의미인지 이해하지 못했습니다.

보여지는 데이터가 다르고 해당 프로퍼티가 다르긴한데...

> 3. dateFormatter의 경우 원래는 MemoListTableViewController에 뒀다가 CoreData에만 사용되는 것이 아니기 때문에 조금 애매하긴 하다고도 생각이 들긴 했지만, DataManager에 두긴 했는데요. 위에 말씀해주신 내용에 대한 이해가 확실히 되지 않은 상태라 '다른 곳에 정의되겠네요'라는 내용을 보고 당장 생각이 들기로는 원래대로 사용하는 곳인 MemoListTableViewController의 extension으로 해당 포맷터를 옮기는 것이 더 적절하지 않을까 생각이 드는데요. 이 것을 의미하신게 맞으실까요? ㅠㅠ

leejun6694 commented on CloudNotes/CloudNotes/Shared/DataManager.swift:
```diff
@@ -0,0 +1,91 @@
+
+    lazy var persistentContainer: NSPersistentCloudKitContainer = {
+        let container = NSPersistentCloudKitContainer(name: "CloudNotes")
+        container.loadPersistentStores(completionHandler: { (_, error) in
```
`(completionHandler:`
생략 가능한 코드로 보입니다ㅎㅎ 참고 부탁드려요~

leejun6694 replied:
`(_, error)` 에서 괄호도...?

Hosinging replied:
1. persistentContainer 클로저 실행코드는 AppDelegate의 코어데이터 스택 초기화 부분을 그대로 가져와서 사용한 것인데요. loadPersistentStores(completionHandler:) 메서드를 제가 정의한 것이 아니기 때문에 기본값을 줘서 생략한다든지, 파라미터를 지운다든지 할 수가 없었습니다..ㅠㅠ


2. (_, error)에서 괄호는 생략 처리했습니다.

leejun6694 commented:
step 2 도 구현하느라 고생 많으셨습니다 : )
좀 더 생각해봤으면 하는 코드들에 코멘트를 남겼습니다. 참고 부탁드릴게요~!!

> 1. 자잘자잘한 Extension 추가

저도 그렇고 현업에서도 extension 을 많이 사용합니다. 다만, 협업하는 동료가 많고, 프로젝트가 크다보니 범용적으로 쓰이지 않는다면 함부로 추가를 하면 안되겠죠ㅎㅎ
제가 공부하던 시점에서는 `이 기능을 위한 라이브러리를 만든다면, 이 코드가 필요할까?` 정도를 개인적인 기준으로 삼았던 것 같습니다.
참고 부탁드려요~~!
지금 추가해주신 코드들도 잘 만들어주셨습니다 👍 

> 2. 노티피케이션 추가

코드 리뷰로 남겼습니다 : )

Hosinging committed:
feat: 리뷰 내용 반영

Hosinging committed:
refactor: MemoListTableViewController내 코드 리팩터링
- MemoListBranching.swift파일 내 MemoListBranching 프로토콜 선언 및 요구사항 정의
- SampleData.swift파일의 SampleData타입에 MemoListBranching프로토콜 채택 및 구현
- DataManager.swfit파일에 MemoEntity를 확장하는 코드 추가
- MemoListTableViewController의 dataSource, delegate메서드 내에 연산 프로퍼티로 메모 데이터 분기해서 표시하는 코드 추가
